openshift/machine-config-operator
# [OCPNODE-553] Add CRD ImageDigestMirrorSet,ImageTagMirrorSet

**- What I did**
- Add support for ImageDigestMirrorSet,ImageDigestMirrorSet mirror configuration CRDs.
- Reject the update of one of current icsp and new idms.itms resources if both exist to avoid having both icsp and new CRs on the same cluster.

Close card: https://issues.redhat.com/browse/OCPNODE-553
Epic: https://issues.redhat.com/browse/OCPNODE-521 
Dependent PR: https://github.com/openshift/runtime-utils/pull/15

**- How to verify it**

Apply following CR, check the /etc/containers/registries.conf

```
apiVersion: config.openshift.io/v1
kind: ImageTagMirrorSet
metadata:
  name: tag-mirror
spec:
  imageTagMirrors:
  - mirrors:
    - example.io/example/ubi-minimal 
    source: registry.access.redhat.com/ubi8/ubi-minimal
  - mirrors:
    - example.io/example/ubi-minimal 
    source: registry.access.redhat.com/ubi8/ubi-minimal-1
    mirrorSourcePolicy: NeverContactSource
```
/etc/containers/registries.conf
```
unqualified-search-registries = ["registry.access.redhat.com", "docker.io"]
short-name-mode = ""

[[registry]]
  prefix = ""
  location = "registry.access.redhat.com/ubi8/ubi-minimal"

  [[registry.mirror]]
    location = "example.io/example/ubi-minimal"
    pull-from-mirror = "tag-only"

[[registry]]
  prefix = ""
  location = "registry.access.redhat.com/ubi8/ubi-minimal-1"
  blocked = true

  [[registry.mirror]]
    location = "example.io/example/ubi-minimal"
    pull-from-mirror = "tag-only"
```

**- Description for the changelog**

Add support for ImageDigestMirrorSet,ImageDigestMirrorSet CRDs
